src/main/java/build/buf/protovalidate/Uri.java

```diff
@@ -0,0 +1,959 @@
+
+    // If an error occurred, return false as invalid.
+    if (f.isError()) {
+      return false;
+    }
+    // Flush the buffer
+    f = this.utf8Decoder.flush(out);
+
+    // If an error occurred, return false as invalid.
+    // Otherwise return true.
+    return !f.isError();
+  }
+
+  /**
+   * Reports whether string contains a valid host.
+   *
+   * <p>Method parses the rule:
+   *
+   * <pre>host = IP-literal / IPv4address / reg-name.
+   */
+  private boolean host() {
+    if (this.index >= this.str.length()) {
+      return true;
+    }
+
+    int start = this.index;
+    this.pctEncodedFound = false;
+
+    // Note: IPv4address is a subset of reg-name
+    if ((this.str.charAt(this.index) == '[' && this.ipLiteral()) || this.regName()) {
+      if (this.pctEncodedFound) {
+        String rawHost = this.str.substring(start, this.index);
+        // RFC 3986:
+        // > URI producing applications must not use percent-encoding in host
+        // > unless it is used to represent a UTF-8 character sequence.
+        if (!this.checkHostPctEncoded(rawHost)) {
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Reports whether string contains a valid port.
+   *
+   * <p>Method parses the rule:
+   *
+   * <pre>port = *DIGIT
+   *
+   * Terminated by end of authority.
+   */
+  private boolean port() {
+    int start = this.index;
+
+    while (true) {
+      if (this.digit()) {
+        continue;
+      }
+
+      if (this.isAuthorityEnd()) {
+        return true;
+      }
+
+      this.index = start;
+
+      return false;
+    }
+  }
+
+  /**
+   * Reports whether string contains a valid IP literal.
+   *
+   * <p>Method parses the rule from RFC 6874:
+   *
+   * <pre>IP-literal = "[" ( IPv6address / IPv6addrz / IPvFuture  ) "]"
+   */
+  private boolean ipLiteral() {
+    int start = this.index;
+
+    if (this.take('[')) {
+      int j = this.index;
+
+      if (this.ipv6Address() && this.take(']')) {
+        return true;
+      }
+
+      this.index = j;
+
+      if (this.ipv6Addrz() && this.take(']')) {
+        return true;
+      }
+
+      this.index = j;
+
+      if (this.ipvFuture() && this.take(']')) {
+        return true;
+      }
+    }
+
+    this.index = start;
+
+    return false;
+  }
+
+  /**
+   * Reports whether string contains a valid ipv6 address.
+   *
+   * <p>ipv6Address parses the rule "IPv6address".
+   *
+   * <p>Relies on the implementation of isIp.
+   */
+  private boolean ipv6Address() {
+    int start = this.index;
+
+    while (this.hexDig() || this.take(':')) {}
+
+    if (CustomOverload.isIp(this.str.substring(start, this.index), 6)) {
+      return true;
+    }
+
+    this.index = start;
+
+    return false;
+  }
+
+  /**
+   * Reports whether string contains a valid IPv6addrz.
+   *
+   * Method parses the rule:
+   *
+   * <pre>IPv6addrz = IPv6address "%25" ZoneID
+   */
```

Comment:
The methods are direct implementations of the ABNF rules. The comment on the reference implementation is:

```
  // Parses the rule from RFC 6874:
  //
  //     IPv6addrz = IPv6address "%25" ZoneID
```

This provides two important bits of information:
- The rule we're implementing. 
- Where it's defined, if not implied by preceding comments.

We have added summarizing sentences in Go and Python to follow best practices, but do we need them here? It's not clear what `string` refers to, and "contains" is misleading.

So I suggest:

```suggestion
  /**
   * Parses the rule from RFC 6874:
   *
   * <pre>IPv6addrz = IPv6address "%25" ZoneID
   */
```

Similar for all other methods.

Reply:
Agreed: https://github.com/bufbuild/protovalidate-python/pull/277